File: python-algorithms/algorithms/sorting/selection_sort.py
from typing import List, TypeVar

DataT = TypeVar("DataT")


def selection_sort(arr: List[DataT]):
    length = len(arr)
    for i in range(length):
        smallest = arr[i]
        for j in range(i + 1, length):
            if arr[smallest] > arr[j]:
                smallest = j
        arr[smallest], arr[i] = arr[i], arr[smallest]

File: rust-algorithms/src/sorting/selection_sort.rs
pub fn selection_sort<T: Ord>(arr: &mut [T]) {
    for i in 0..arr.len() {
        let mut smallest = i;
        for j in i..arr.len() {
            if arr[j] < arr[smallest] {
                smallest = j;
            }
        }
        arr.swap(i, smallest);
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_selection_sort() {
        let mut res = vec!["d", "a", "c", "b"];
        selection_sort(&mut res);
        assert_eq!(res, vec!["a", "b", "c", "d"]);
    }
}
